src/Rings/mpolyquo-localizations.jl

```
@@ -2699,27 +2699,50 @@ function Base.:(==)(
 
   if _has_coefficient_map(f)
     if _has_coefficient_map(g) 
-      coefficient_map(f) == coefficient_map(g) || return false
+      if coefficient_map(f) isa Map && coefficient_map(g) isa Map
+        coefficient_map(f) == coefficient_map(g) || return false
+      else
+        error("comparison of `coefficient_map`s of type $(typeof(coefficient_map(f))) and $(typeof(coefficient_map(g))) is not implemented")
+      end
     else
-      coefficient_map(f) == identity_map(coefficient_ring(domain(f))) || return false
+      return is_trivial(coefficient_map(f))
     end
   elseif _has_coefficient_map(g)
     if _has_coefficient_map(f) 
-      coefficient_map(f) == coefficient_map(g) || return false
+      if coefficient_map(f) isa Map && coefficient_map(g) isa Map
+        coefficient_map(f) == coefficient_map(g) || return false
+      else
+        error("comparison of `coefficient_map`s of type $(typeof(coefficient_map(f))) and $(typeof(coefficient_map(g))) is not implemented")
+      end
     else
-      coefficient_map(g) == identity_map(coefficient_ring(domain(g))) || return false
+      is_trivial(coefficient_map(g))
     end
   end
 
-  return all(x->f(x) == g(x), gens(domain(f)))
+  return all(f(x) == g(x) for x in gens(domain(f)))
+end
+
+# TODO: Move to wherever `MapFromFunc` comes from
+function (==)(f::MapFromFunc, g::MapFromFunc)
+  f === g && return true
+  domain(f) === domain(g) || return false
+  codomain(f) === codomain(g) || return false
+  f.f === g.f && return true
+  error("comparison of maps $f and $g not possible")
+end
```

This does not look like a great solution for me, since it most cases the `.f` fields will just be anonymous functions that are not comparable.
As such, I would find it more natural for MapFromFunc to have no custom `==` at all, thus falling back to `===`

---

> As such, I would find it more natural for MapFromFunc to have no custom `==` at all, thus falling back to `===`

We could do that, yes. But in that case it is not possible to require `==` to have a mathematically correct implementation for `Map`s in general. I.e. we can not restrict to cases with some type-parameter being `<:Map` and then safely assume that `==` can be used in the body of the function. 

To me `Map` seemed like a good cutting point to distinguish cases which can be handled from those which can't. But with your suggestion, I'm afraid this would fall.

---

I think we had this discussion already about 2 years ago, but unfortunately I cannot find it anymore.
What I think we concluded to (but this could also be totally wrong):
`==` can in many cases not be any mathematical equivalence. Because what should e.g. `G == H` for two groups do? Check if they are isomorphic? That's way more complicated than what we usually want `==` to be.
Thus, we use `==` more as a object equality, i.e. the two objects are somehow defined in the same way and thus seem to be equivalent. This is, in general, nothing *mathematically sane* to reason about.

But I think this here is such an occurrence where we just *cannot* do something mathematically sane, since anonymous function can do too many things.

---

@lgoettgens is your point that maps should be compared like groups? I don't remember that we agreed on this.

If map/homomorphism comparison should be doing something mathematical, then the MapFromFunc should compare equal if they are egal (`===`) and it should throw an error otherwise.

---

Ok. It would be good if we could dig out the conclusions of that discussion again. I agree that there are cases where we can not reasonably check mathematical equality with generic code. Hence, the mathematically rigorous solution would be to throw an error in these cases (I think). 

In general, one can, of course, communicate to the user that e.g. for groups `==` is not implemented in the mathematical way. But in this case here (morphisms of modules), one would probably expect the mathematically rigorous answer. 

Another possible solution that I see is to not allow `==` for morphisms of modules with a `ring_map` at all (and probably similar for morphisms of `MPolyAnyRing`s?). But then again, I would not write it somewhere in the docstring that "the user should be aware that mathematical comparison via `==` only works in abscence of a `ring_map`". Since that distinction is not tied to the outer type, but only to one of its parameters, the user will probably not find that information. So even there I would implement `==` to throw an error.

---

> the MapFromFunc should compare equal if they are egal (`===`) and it should throw an error otherwise.

I don't object adding this to Hecke.

I am not even sure myself what point I wanted to make above... hmm...

---

Just to make sure. Is something failing here without this `==` for `MapFromFunc`? I might be tempted to make it to always return `false` otherwise (until someone has an application).

---

Yes: The tests as they are now would fail without this implementation. In general, my code is using `MapFromFunc` in many places as maps for coefficients. Not sure how much comparison is done with these objects, in the end, though. 

If there are good reasons for always returning `false`, then I would have to revise this. But otherwise, I would prefer to keep it like this.